reconcile/external_resources/manager.py

```diff
@@ -298,47 +298,50 @@ def _get_reconciliation_status(
                     r.key,
                 )
                 if r.action == Action.APPLY:
-                    status.resource_status = ResourceStatus.PENDING_SECRET_SYNC
+                    reconciliation_status.resource_status = (
+                        ResourceStatus.PENDING_SECRET_SYNC
+                    )
                 elif r.action == Action.DESTROY:
-                    status.resource_status = ResourceStatus.DELETED
-                status.reconcile_time = (
+                    reconciliation_status.resource_status = ResourceStatus.DELETED
+                reconciliation_status.reconcile_time = (
                     self.reconciler.get_resource_reconcile_duration(r) or 0
                 )
+                reconciliation_status.needs_state_update = True
             case ReconcileStatus.ERROR:
                 logging.info(
                     "Reconciliation ended with ERROR: Action:%s, Key:%s",
                     r.action.value,
                     r.key,
                 )
-                status.resource_status = ResourceStatus.ERROR
+                reconciliation_status.resource_status = ResourceStatus.ERROR
+                reconciliation_status.needs_state_update = True
             case ReconcileStatus.NOT_EXISTS:
                 logging.info(
                     "Reconciliation should exist but it doesn't. Marking as ERROR to retrigger: Action:%s, Key:%s",
                     r.action.value,
                     r.key,
                 )
-                status.resource_status = ResourceStatus.ERROR
+                reconciliation_status.resource_status = ResourceStatus.ERROR
+                reconciliation_status.needs_state_update = True
+            case ReconcileStatus.IN_PROGRESS:
+                logging.debug("Reconciliation still in progress ...")
 
-        return status
+        return reconciliation_status
 
     def _update_resource_state(
         self,
         r: Reconciliation,
         state: ExternalResourceState,
         reconciliation_status: ReconciliationStatus,
     ) -> None:
-        if (
-            state.resource_status.is_in_progress
-            and reconciliation_status.resource_status.is_in_progress
-        ):
-            logging.debug(
-                "Reconciliation is still in progress. There is no need to update the state"
-            )
+        if not reconciliation_status.needs_state_update:
+            logging.debug("Reonciliation does not need to update the resource state")
             return
-        state.ts = datetime.now(UTC)
+
         if reconciliation_status.resource_status == ResourceStatus.DELETED:
             self.state_mgr.del_external_resource_state(r.key)
         else:
+            state.ts = datetime.now(UTC)
```

Comment:
When `ts` should change? Should it only change when `state.resource_status` changed?

Reply:
Not exactly. Whenever it changes from in-progress to another state.

Reply:
then you can manage it inside `state`, instead of

```python
            state.ts = datetime.now(UTC)
            state.resource_status = reconciliation_status.resource_status
```

the can be done as a state method


```python
state.update_status(reconciliation_status.resource_status)
```

and handles it internally

```python
def update_status(self, status):
    if self.resource_status == in_progress and status != in_progress:
        self.ts = datetime.now(UTC)
    self.resource_status = status
```

Reply:
OK. I moved some things around to do this and make things a bit clearer (at least for me). PTAL cc: @chassing